### DISCUSSION TITLE
Usage of `wezterm ssh` with ssh-agent on Windows 11

### DISCUSSION THREAD

**Discussion by antoineco** [wezterm/wezterm#3772] [Q&A] [answered]
I have the "OpenSSH Authentication Agent" (ssh-agent) service running on my machine, and I successfully added my SSH key using `ssh-add`:

(PowerShell 7.3.4)

```console
> ssh-add .\.ssh\auth
Enter passphrase for .\.ssh\auth:
Identity added: .\.ssh\auth (acotten)
```

```console
> ssh-add -L
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIBrWkMRdF5phpLftdUHIgcSnSIKunqBecVN1jgSkuz8H acotten
```

However, `wezterm ssh` is not picking the key from the agent, and prompts me to enter the passphrase again:

```console
> wezterm ssh myhost
(new window)
Using libssh-rs to connect to acotten@XXX.XXX.XXX.XXX:22
SSH-2.0-OpenSSH_8.4p1 Debian-5+deb11u1
Passphrase (C:\Users\acott\.ssh\auth):
```

This does not happen if I connect to the same host with the `ssh` command.

Could anyone share some pointers about how to solve this?

---

**Update**

After [some research](https://developer.1password.com/docs/ssh/agent/compatibility/#openssh), I found that the OpenSSH agent's (named) pipe was located at `\\.\pipe\openssh-ssh-agent`.

I tried adding the following directive to my configuration:

```ssh_config
Host bench
  User acotten
  Hostname XXX.XXX.XXX.XXX
  IdentityFile C:\Users\acott\.ssh\auth
  IdentitiesOnly true
  IdentityAgent \\.\pipe\openssh-ssh-agent
```

Unfortunately this doesn't seem supported by `libssh-rs`:

```
INFO   libssh::ssh_socket_unix                    > Error from connect(\\.\pipe\openssh-ssh-agent): No such file or directory
INFO   libssh::ssh_userauth_publickey_auto        > Trying to authenticate with C:\Users\acott\.ssh\auth
```

**Comment by wez** [comment]
Is it possible to configure your ssh agent to use a unix domain socket instead of a named pipe?

**Reply by antoineco** [reply]
I did not even know this was a thing on Windows, until I read https://devblogs.microsoft.com/commandline/af_unix-comes-to-windows/ 👀 

The answer is no, but your comment prompted me to search Microsoft's related OSS project, and there is a PR open to support that: https://github.com/PowerShell/openssh-portable/pull/674

**Reply by wez** [reply, marked as answer]
Ah, good. I was the person that added IdentityAgent support to libssh, and enabled unix domain support there for Windows (at my former place of work, we had a custom ssh agent that did some stuff with the TPM to manage secrets, and that was exposed via unix domain).

I commented on another question/issue from you about this or a similar thing: you'll need libssh to support named pipes in order for wezterm to support them.

Making unix software support named pipes is generally difficult because the semantics are similar in some ways but very different in some very painful other ways; named pipes can block forever in a number of situations, which is usually at odds with the non-blocking approach used in the vast majority of unix applications.

I personally won't be working to enable named pipes in libssh because I have no need for it.

I think the best longer term bet is MS shipping unix domain support.

In the shorter term, if I were in your situation I would consider:

* Is there an alternative ssh agent that I could use?  eg: using WSL 1 you can use interop to have it place the socket on the host filesystem and share out from the agent inside a WSL container
* Could you do something netcat-or-socat-like with a program that listens on a unix socket and proxies the connections to the named pipe?  That program probably wouldn't be all that difficult to write if there isn't already a thing that does it

**Reply by wez** [reply]
this comment might point you to the second option I suggested:
* https://github.com/wez/wezterm/issues/3704#issuecomment-1542876416

**Reply by antoineco** [reply]
Those are good suggestions.

I agree that supporting named pipes isn't a requirement since `libssh` already supports Unix domain sockets, and the fact that Windows itself supports AF_UNIX changes everything.

[npiperelay](https://github.com/jstarks/npiperelay) indeed looks simple enough. I will look into using it in conjunction with [ncat](https://nmap.org/book/ncat-man.html#ncat-man-unixsock), for the Unix socket end on Windows.

---

WSL2 side notes:

- I attempted to configure the client to connect directly to the SSH agent running inside my main WSL instance over the 9p share, but this isn't supported:

    ```ssh_config
    Host bench
      IdentityAgent \\wsl.localhost\Ubuntu\tmp\ssh-XXXXXXviy435\agent.397
    ```

- Starting the SSH agent within WSL2 with a bind address on the host's 9p share is also not supported:

    ```console
    $ ssh-agent -a /mnt/c/Users/acotten/.ssh/agent.sock
    unix_listener: cannot bind to path /mnt/c/Users/acotten/.ssh/agent.sock: Operation not supported
    ```

**Reply by naikrovek** [reply]
> at my former place of work, we had a custom ssh agent that did some stuff with the TPM to manage secrets, and that was exposed via unix domain

😲

**Reply by GottZ** [reply]
I just noticed `IdentitiesOnly yes` disables agent support completely inside the ssh2 backend so.. #7739 here ye go.

**Comment by Oddegamra** [comment]
Having struggled with this problem myself today, I will add for anyone happening across this page that setting `ssh_backend = "Ssh2"`  in Wezterm's configuration will, in fact, connect to the OpenSSH Agent on Windows via named pipe if it is set up correctly in `~/.ssh/config`. This also applies to Putty's Pageant, which now also uses named pipes.

**Reply by rgwood** [reply]
Great tip!

Confirmed that `config.ssh_backend = "Ssh2"` uses the OpenSSH Agent as expected. I did not need to make any changes to my `~/.ssh/config` file.

**Reply by acuteaura** [reply]
"unable to exchange encryption keys" for me when using 1password, which you also can't configure to use a unix socket.

**Reply by gsommer** [reply]
It does not work for me. What do you mean with "if it is set up correctly in ~/.ssh/config"?

If I use ssh in a powershell window, I do not need to enter the password. but using "wezterm ssh" prompts for the password, though wezterm uses the configuration from .ssh\config, same as the ssh command (I did not put any configuration into wezterm.lua).

It also works from Windows Terminal.

I would really appreciate your help, as always having to enter the password is annoying.

**Reply by michaelrommel** [reply]
@gsommer : You wrote:

> I did not put any configuration into wezterm.lua

but you should put `config.ssh_backend = "Ssh2"` in the `.wezterm.lua` config file.
In the config file under `C:\Users\<username>\.ssh\config` I have:
```
IdentityFile ~/.ssh/id_ed25519
ForwardAgent yes
ForwardX11 true
ForwardX11Trusted false
ServerAliveInterval 25

Host WSL
HostkeyAlias localhost
Hostname 127.0.0.1
User <username>
```
With that I can connect to the domain `SSH:WSL` without being asked for a password.

**Comment by t-mart** [comment]
I found this page looking for help on why Window's 1Password SSH agent was not working for *only* WezTerm with normal shell invocations of **`ssh <host>`**. Other terminals and applications would invoke the agent just fine. (In other words, my issue was not with `wezterm ssh`. And, I apologize for hopping on this different Q&A, but googling led me here and maybe others).

Here's how I fixed it:

```lua
if wezterm.target_triple == 'x86_64-pc-windows-msvc' then
  config.default_ssh_auth_sock = "\\\\.\\pipe\\openssh-ssh-agent"
end
```

(I add the [`target_triple`](https://wezterm.org/config/lua/wezterm/target_triple.html) condition so that my config works cross-platform. If you're Windows-only, you won't need it.)

I don't really understand these pipes/socks/etc, but WezTerm sets an environment variable on spawned shells: `SSH_AUTH_SOCK` to something in `~/.local/share/wezterm`. This value is controlled by [`config.default_ssh_auth_sock`](https://wezterm.org/config/lua/config/default_ssh_auth_sock.html). But, the [1Password SSH agent expects](https://developer.1password.com/docs/ssh/get-started/#step-3-turn-on-the-1password-ssh-agent) to be communicated with at `\\.\pipe\openssh-ssh-agent`. So, therefore, the 1Password agent is not invoked as long as WezTerm is changing this, and above, we change it "back".

It would be nice if there would be a way to not have WezTerm set any environment variable here -- in other terminals, the absence of `SSH_AUTH_SOCK` let the 1Password SSH agent just work. Ideally, setting `config.default_ssh_auth_sock` to `nil` would behave this way, but on nightly, doing so seemed to be ignored and overwritten.

**Reply by justinbarclay** [reply]
Thanks so much for this, I just installed nightly and 1password's SSH integration stopped working.

**Reply by Kasheen** [reply]
For some reason this doesn't work as expected for me.

I'm using wezterm with nushell and `echo $env.SSH_AUTH_SOCK` still produces a path inside of `~/.local/share`.

Yet using `wezterm cli list-clients` shows the correctly set named pipe as defined in the config.

Not really sure why this is happening and I'm having to set `SSH_AUTH_SOCK` manually in nushell.

Edit: It's also happening with CMD, and I've also tried removing the guard around setting the config variable.

**Comment by AM-I-Human** [comment]
I found a similar issue.
I upgraded on nightly and I saw that my ssh-agent for the credentials, wasn't working anymore.
I am on Windows, no WSL.
I found this curious error:
```
❯ ssh-add
Error connecting to agent: No such file or directory
```
That suggest me that there is something wired wrongly on the nightly, but not on the latest.
I think is related, so I added this comment, maybe it gives some clues on the problem

**Reply by kmoschcau** [reply]
I had this problem as well. But it seems just a Windows restart later and the problem is mysteriously gone.

**Reply by kmoschcau** [reply]
And it's back. So no idea what made it work for some time.

**Reply by kmoschcau** [reply]
I found a workaround for now: You can just unset the `SSH_AUTH_SOCK` environment variable. For example like so in Powershell:

```ps1
$env:SSH_AUTH_SOCK = $null
```

**Reply by cco** [reply]
> I found a workaround for now: You can just unset the `SSH_AUTH_SOCK` environment variable. For example like so in Powershell:
> 
> ```powershell
> $env:SSH_AUTH_SOCK = $null
> ```

Does setting ``config.default_ssh_auth_sock = "\\\\.\\pipe\\openssh-ssh-agent"`` in your config not work for you?

**Reply by kmoschcau** [reply]
@cco I swear I have tried that before, but probably never with a full wezterm restart. It does work when I set `config.default_ssh_auth_sock = "\\\\.\\pipe\\openssh-ssh-agent"`, but only after a full restart.

What you will then notice, when you `echo $env:SSH_AUTH_SOCK`, is that you still get the same kind of files in there as well. However, looking in `~\.local\share\wezterm`, wezterm now also creates a symlink called `agent.<number>` in addition to the previously created `gui-sock-<number>`. That symlink seems like it was the missing puzzle piece to actually make it work.

**Reply by cco** [reply]
> @cco I swear I have tried that before, but probably never with a full wezterm restart. It does work when I set `config.default_ssh_auth_sock = "\\\\.\\pipe\\openssh-ssh-agent"`, but only after a full restart.

Yeah, I think this added an extra chunk of time attempting to debug the issue. With regular shells, you generally expect .bashrc and other env updates to take effect immediately in a new shell session, but Wezterm is holding onto the socket configuration until a full restart it looks like.

**Comment by bltavares** [comment]
For anyone on Windows 11 reading this in 2026 (Nightly `wezterm version: 20260622-120102-6ff54928 x86_64-pc-windows-msvc`) here is how I integrated `ssh-agent.exe`, `ssh.exe` and Wezterm SSH Domains.

The author's selected answer had a lot of good explanations, but it took me a while to understand how `SSH_AUTH_SOCK`, Wezterm, `~/.ssh/config` and `OpenSSH Agent` service was linked together.

Windows' builtin `ssh.exe` and `ssh-agent.exe` communicate with the Windows named socket on `\\.\pipe\openssh-ssh-agent`, but Wezterm requires a Unix socket path for integration. Even tho there is a [6 years](https://github.com/PowerShell/Win32-OpenSSH/issues/1024) open issue with backlog tag and a few proposed PRs to include a Unix Socket integration, Windows users will require some workarounds to bridge the gap. I figured out a couple of options for now:

1. Run a bridge between Windows Named Socket and Unix Socket

This is suggested by the author's accepted answer, using `socat` and `npiperelay.exe` to create a Unix Socket file interop with the Windows OpenSSH Agent named pipe. There are several other helper programs to fill in this scope (not recommendations, analyze and choose your preferred option)

- Powershell/bat script with `socat.exe` and `npiperelay.exe`
- https://github.com/benpye/wsl-ssh-pageant
- https://github.com/rupor-github/wsl-ssh-agent
- https://github.com/benpye/wsl-ssh-pageant
- https://github.com/Race666/wsl2-to-pageant-forwarder
- https://github.com/mame/wsl2-ssh-agent
- https://github.com/cuviper/ssh-pageant
- https://github.com/jet2jet/stream-connector

Some helpers come with a tray icon, others have WSL helpers, etc. To integrate with Wezterm it only need to create a Unix Socket on an accessible location.

2. Choose an alternative SSH Agent program

The Windows builtin SSH Agent service only exposes the named socket, and others on this thread mentioned a few other Windows programs. As of 2026, it seems a few options are possible (not recomendations, analyze and choose your preferred option):

- KeeAgent plugin for Keepass
- `peagent.exe --unix` flag from Putty
- https://github.com/masahide/OmniSSHAgent (can also work as a Proxy mode)

The Unix socket is often labeled `WSL1 Socket` integration. (1Password Agent seems to also only use named pipes in Windows, so it would also need one of the bridge options).

As long as there is a Unix Socket somewhere, you can integrate Wezterm and `ssh.exe`. I've configured as the following.

1. `~/.ssh/config` - Enable forwarding but don't add an `IdentityAgent` config

If there is no `IdentityAgent` configurd, `ssh.exe` will use the default Windows named pipe on `\\.\pipe\openssh-ssh-agent`. Adding `ForwardAgent` signals both Wezterm and `ssh.exe` it should forward identities.

```conf
# ~/.ssh/config
Host myhost.internal
  User myuser
  ForwardAgent yes
```

2. Avoid Wezterm `SSH_AUTH_SOCK` variable

The `mux_enable_ssh_agent=true` will populate the env var with a Wezterm managed symlink on `.local/share/wezterm`. When this variable is set, it will break `ssh.exe` as it will attempt to use the broken symlink instead of the default named socket.

```lua
config.mux_enable_ssh_agent = false
```

3. Point Wezterm SSH domains to the Unix Socket
Configure Wezerm to use the Unix socket from either a bridge method or a SSH Agent with support.

```lua
config.ssh_domains = wezterm.default_ssh_domains()
for _, dom in ipairs(config.ssh_domains) do
  -- Adjust the socket path to your preferred location
  dom.ssh_option.identityagent = string.format('%s\\ssh-agent.sock', os.getenv('USERPROFILE'))
end
```

This will enable `wezterm cli spawn --domain SSH:myhost.internal` or the shortcuts for `wezterm.action.SpawnTab` to integrate with the Unix socket, exposing the credentials to the host. Using `config.default_ssh_auth_sock` will populate the `$ENV:SSH_AUTH_SOCK` on `Local` domains and break `ssh.exe` integration.

The `ForwardAgent yes` config will be read from the `~/.ssh/config` and the `IdentityAgent` path will be defined in the config. This way, `ssh.exe` will use the default named pipe while Wezterm will use the Unix socket instead.

To use `wezterm ssh`, pass in the socket as an additional option:

```pwsh
wezterm ssh gibson.internal -o identityagent=$ENV:USERPROFILE/ssh-agent.sock
```

Putting all together, here is my setup:

- `~/.ssh/config`:

```conf
Host myhost.internal
  User myuser
  ForwardAgent yes
```

- `~/.config/wezterm/wezterm.lua`

```lua
local wezterm = require("wezterm")
local config = wezterm.config_builder()
if wezterm.target_triple:find("windows") ~= nil then
  config.default_prog = { "pwsh", "-NoLogo" }
  config.keys = {
    { key = "!", mods = "CTRL|SHIFT", action = wezterm.action.SpawnTab({ DomainName = "local" }) },
    { key = "#", mods = "CTRL|SHIFT", action = wezterm.action.SpawnTab({ DomainName = "SSH:myhost.internal" }) },
  }
  --
  -- Remove wezterm symlinking populating SSH_AUTH_SOCK from env to make ssh.exe use default windows named socket
  config.mux_enable_ssh_agent = false
  config.ssh_domains = wezterm.default_ssh_domains()
  for _, dom in ipairs(config.ssh_domains) do
    -- point wezterm internal ssh session to ssh-agent Unix socket at $HOME
    dom.ssh_option.identityagent = string.format('%s\\ssh-socket.sock', os.getenv('USERPROFILE'))
  end
end
```
